cmd/rocsp-tool/main.go

```diff
@@ -314,23 +360,43 @@ func (cl *client) loadFromDB(ctx context.Context, speed ProcessingSpeed) error {
 	rowTicker := time.NewTicker(frequency)
 
 	statusesToSign := make(chan *sa.CertStatusMetadata)
-	successes := make(chan struct{}, speed.ParallelSigns)
+	successes := make(chan uint64, speed.ParallelSigns)
 	errChan := make(chan error, speed.ParallelSigns)
 
+	// a set of all inflight certificate statuses, indexed by their `ID`.
+	inflightIDs := newInflight()
+
 	var signersWg sync.WaitGroup
 
 	for i := 0; i < speed.ParallelSigns; i++ {
 		signersWg.Add(1)
 		go cl.signAndStoreResponses(statusesToSign, errChan, successes, &signersWg)
 	}
-	var n, errorCount int
-	// Consume all available successes and errors from the output channels.
-	// If none are immediately available, return.
-	consume := func() {
+	var successCount, errorCount int
+	var scanned int
+	for rows.Next() {
+		<-rowTicker.C
+
+		status := new(sa.CertStatusMetadata)
+		if err := sa.ScanCertStatusMetadataRow(rows, status); err != nil {
+			return fmt.Errorf("scanning row %d: %w", successCount, err)
+		}
+		scanned++
+		inflightIDs.add(uint64(status.ID))
+		if scanned%100000 == 0 {
+			log.Printf("scanned %d certificateStatus rows. minimum inflight ID %d", scanned, inflightIDs.min())
+		}
+		statusesToSign <- status
+
+		// Consume all available successes and errors from the output channels.
+		// If none are immediately available, return.
+		consume:
 		for {
 			select {
-			case <-successes:
-				n++
+			case doneID :=<-successes:
+				inflightIDs.remove(doneID)
```

Comment:
Slightly unfortunate that we only remove inflight IDs on success, not on error. Of course, one that errored out hasn't been inserted, so we might want to go clean it up later. But it also means that a single early failure causes us to have no insight into what the later progress was like.

Reply:
Good point. I will make errors also remove the inflight ID. That will support restarting at a given ID, and to deal with the fact that some fraction has errored, we can always do another pass through.